yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ConfigurableTokenManager.scala

```diff
@@ -0,0 +1,214 @@
+ * required for this service.
+ * 2. set spark.yarn.security.tokens.{service}.enabled to true
+ * 3. Specify the class name through spark.yarn.security.tokens.{service}.class
+ *
+ */
+final class ConfigurableTokenManager private[yarn] (sparkConf: SparkConf) extends Logging {
+  private val tokenProviderEnabledConfig = "spark\\.yarn\\.security\\.tokens\\.(.+)\\.enabled".r
+  private val tokenProviderClsConfig = "spark.yarn.security.tokens.%s.class"
+
+  // Maintain all the registered token providers
+  private val tokenProviders = mutable.HashMap[String, ServiceTokenProvider]()
+
+  private val defaultTokenProviders = Map(
+    "hdfs" -> "org.apache.spark.deploy.yarn.token.HDFSTokenProvider",
+    "hive" -> "org.apache.spark.deploy.yarn.token.HiveTokenProvider",
+    "hbase" -> "org.apache.spark.deploy.yarn.token.HBaseTokenProvider"
+  )
+
+  // AMDelegationTokenRenewer, this will only be create and started in the AM
+  private var _delegationTokenRenewer: AMDelegationTokenRenewer = null
+
+  // ExecutorDelegationTokenUpdater, this will only be created and started in the driver and
+  // executor side.
+  private var _delegationTokenUpdater: ExecutorDelegationTokenUpdater = null
+
+  def initialize(): Unit = {
+    // Copy SparkConf and add default enabled token provider configurations to SparkConf.
+    val clonedConf = sparkConf.clone
+    defaultTokenProviders.keys.foreach { key =>
+      clonedConf.setIfMissing(s"spark.yarn.security.tokens.$key.enabled", "true")
+    }
+
+    // Instantialize all the service token providers according to the configurations.
+    clonedConf.getAll.filter { case (key, value) =>
+      if (tokenProviderEnabledConfig.findPrefixOf(key).isDefined) {
+        value.toBoolean
+      } else {
+        false
+      }
+    }.map { case (key, _) =>
+      val tokenProviderEnabledConfig(service) = key
+      val cls = sparkConf.getOption(tokenProviderClsConfig.format(service))
+        .orElse(defaultTokenProviders.get(service))
+      (service, cls)
+    }.foreach { case (service, cls) =>
+      if (cls.isDefined) {
+        try {
+          val tokenProvider =
+            Utils.classForName(cls.get).newInstance().asInstanceOf[ServiceTokenProvider]
+          tokenProviders += (service -> tokenProvider)
+        } catch {
+          case NonFatal(e) =>
+            logWarning(s"Fail to instantiate class ${cls.get}", e)
+        }
+      }
+    }
+  }
+
+  /**
+   * Get service token provider by name.
+   */
+  def getServiceTokenProvider(service: String): Option[ServiceTokenProvider] = {
+    tokenProviders.get(service)
+  }
+
+  /**
+   * Obtain tokens from all the token providers and add into credentials, also return as an array.
+   */
+  def obtainTokens(conf: Configuration, creds: Credentials): Array[Token[_]] = {
+    val tokenBuf = mutable.ArrayBuffer[Token[_]]()
+    tokenProviders.values.foreach { provider =>
+      if (provider.isTokenRequired(conf)) {
```

Comment:
Seems like you could remote providers that don't require tokens during `initialize` if you make a `SparkConf` available to it. Then you could avoid passing `SparkConf` to all of these methods.

You could also make `Configuration` a constructor args for the same reasons.